### PR TITLE
remove bank epoch_scheduler. use rent_collector's

### PR DESCRIPTION
**Pull request by apfitzge**
#### Problem
https://discord.com/channels/428295358100013066/439194979856809985/967620591974772756

bank.epoch_schedule and bank.rent_collector.epoch_schedule did not match when loading from a devnet snapshot

#### Summary of Changes
Remove bank.epoch_schedule and use bank.rent_collector.epoch_schdule as the epoch_schedule internally in bank

Fixes #

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -1257,9 +1255,6 @@ pub struct Bank {
     /// latest rent collector, knows the epoch
     rent_collector: RentCollector,
 
-    /// initialized from genesis
-    epoch_schedule: EpochSchedule,
```
Comment (mvines) [inline runtime/src/bank.rs:1261]: Pretty sure CI will barf on this change, removing `epoch_schedule` from`Bank` breaks snapshot backwards compatibility in particular.

Comment (mvines) [reply runtime/src/bank.rs:1261]: Instead we could keep the field in but rename it to `_obsolete_epoch_schedule`

Comment (apfitzge) [reply runtime/src/bank.rs:1261]: It was removed from the struct, but the serialization still has the field there. It's unused in deserialization, and we use the genesis_config.epoch_schedule to set the rent_collector.epoch_schedule as well.

So we are still writing snapshots w/ the same format, and can read snapshots with that format as well, but the initialization from the snapshot is done differently.

It's still unclear to me how the genesis_configs and rent_collectors got out of sync on devnet in the first place though.

Comment (mvines) [reply runtime/src/bank.rs:1261]: Ah, ok cool.  Glad I'm wrong :)

> It's still unclear to me how the genesis_configs and rent_collectors got out of sync on devnet in the first place though.

How are they different?

File: runtime/src/rent_collector.rs [hunk 2]
```diff
@@ -77,6 +77,18 @@ impl RentCollector {
         }
     }
 
+    pub fn clone_with_schedule_and_epoch(
```
Comment (mvines) [inline runtime/src/rent_collector.rs:80]:
```suggestion
    pub(crate) fn clone_with_schedule_and_epoch(
```

**Review by mvines — APPROVED** [review]

File: runtime/src/bank.rs (hunk 1 above)
Comment (apfitzge) [reply runtime/src/bank.rs:1261]: The slots_per_epoch on the rent_collector were 8192 in the snapshot, but 432000 off the genesis config. Same for the leader offset.

Comment (mvines) [reply runtime/src/bank.rs:1261]: 8192 is `DEFAULT_DEV_SLOTS_PER_EPOCH`.  Maybe it's been like that forever on devnet. Oh well, I'm pretty confident this was a devnet-only configuration issue somewhere in the past

**Comment by t-nelson** [comment]
any reason we're retain the rent collector one rather than the bank's?  I thought rent collector source was the one that's wrong

**Comment by apfitzge** [comment]
> any reason we're retaining the rent collector one rather than the bank's? I thought rent collector source was the one that's wrong

simplicity mainly. Bank owns rent_collector, and rent_collector.epoch_schedule is public, so it can still access it that way.

However, with this change we deserialize the rent_collector (including whatever epoch_schedule was deserialized), but then replace it with the epoch_schedule from genesis_config.

**Comment by t-nelson** [comment]
> simplicity mainly. Bank owns rent_collector, and rent_collector.epoch_schedule is public, so it can still access it that way.

right, but it's unintuitive to store it in `RentCollector` given the broad utility of `EpochSchedule` and the fact that there doesn't appear to be any reason for `RentCollector` to hold one since it's only used in one method and there's an `EpochSchedule` available within a few frames of every call site.  seems like an, "asked 'can' instead of 'should'" situation

**Comment by jeffwashington** [comment]
> right, but it's unintuitive to store it in `RentCollector` given the broad utility of `EpochSchedule` and the fact that there doesn't appear to be any reason for `RentCollector` to hold one since it's only used in one method and there's an `EpochSchedule` available within a few frames of every call site. seems like an, "asked 'can' instead of 'should'" situation

@t-nelson note that this pr will be reverted to restore the passing of an additional EpochSchedule through the world if we remove epoch_schedule from rent_collector.
#24893

**Comment by jeffwashington** [comment]
@mvines @t-nelson @apfitzge well, in the latest twist of fate, when I dig into this further, bad things are present.
Turns out we need to keep BOTH of these. devnet's rent_collector.epoch_schedule is used to collect rent. It has to match what the rest of the cluster is using.
the bank's epoch_schedule needs to be used to calculate how many slots there are per epoch in order to get the rent collection partitions by pubkey correct.

**Comment by jeffwashington** [comment]
so the easiest solution is to keep things as they are, but pass bank.epoch_schedule separately to the rehash and load modification code when we skip rewrites.
the revert in #24893 is required to get both `EpochSchedule`s

**Comment by jeffwashington** [comment]
we should discuss whether to restart devnet or patch it or do something to cause these epoch schedules to line up so we can get rid of some complexity that is only required due to devnet.

**Comment by apfitzge** [comment]
Closing this PR, as we ended up doing a revert in https://github.com/solana-labs/solana/pull/24893. Jeff found we actually (currently) need both epoch_schedules on devnet. This PR is no longer necessary.

**Comment by t-nelson** [comment]
is there a new issue to discuss how to get devnet's `EpochSchedule`s synchronized so we can dump one?

**Pull request closed without merging**